inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -2319,74 +2322,93 @@ private <T> void filterReplacedBeans(BeanResolutionContext resolutionContext, Co
+                        replacingCandidate.getBeanType(), replacedBeanType, factory.get());
+                } else {
+                    LOG.debug("Bean [{}] replaces existing bean of type [{}]",
+                        replacingCandidate.getBeanType(), replacedBeanType);
+                }
+            }
 
-                        final AnnotationValue<Replaces> replacesAnn = replacingCandidate.getAnnotation(Replaces.class);
-                        Class beanType = replacesAnn.classValue().orElse(getCanonicalBeanType(replacingCandidate));
-                        if (replacesAnn.contains(NAMED_MEMBER)) {
-                            final String qualifier = replacesAnn.stringValue(NAMED_MEMBER).orElse(null);
-                            if (qualifier != null) {
-                                final Optional qualified = Qualifiers.<T>byName(qualifier)
-                                        .qualify(beanType, Stream.of(definition));
-                                if (qualified.isPresent()) {
-                                    return true;
-                                }
-                                continue;
-                            }
-                        }
+            Optional<Class<?>> declaringType = definitionToBeReplaced instanceof BeanDefinition ?
+                ((BeanDefinition<?>) definitionToBeReplaced).getDeclaringType() :
+                Optional.empty();
 
-                        Optional<Class<?>> factory = replacesAnn.classValue("factory");
-                        if (LOG.isDebugEnabled()) {
-                            if (factory.isPresent()) {
-                                LOG.debug("Bean [{}] replaces existing bean of type [{}] in factory type [{}]", replacingCandidate.getBeanType(), beanType, factory.get());
-                            } else {
-                                LOG.debug("Bean [{}] replaces existing bean of type [{}]", replacingCandidate.getBeanType(), beanType);
-                            }
-                        }
-                        if (comparisonFunction == null) {
-                            comparisonFunction = typeMatches(definition, annotationMetadata);
-                        }
-                        if (factory.isPresent() && declaringType.isPresent()) {
-                            if (factory.get() == declaringType.get()) {
-                                if (comparisonFunction.apply(beanType)) {
-                                    return true;
-                                }
-                            }
-                        } else {
-                            if (comparisonFunction.apply(beanType)) {
-                                return true;
-                            }
-                        }
-                    }
+            if (named.isPresent() || qualifier.isPresent()) {
+                if (named.isPresent() && qualifier.isPresent()) {
```

Comment:
Here you check if optional is present and then you continue with optional in your code. Please first check if `named` qualifies then if `qualifier` qualifies etc. Preferably reduce the number of lamdas.